src/Language/PureScript/ModuleDependencies.hs

```diff
@@ -1,99 +1,65 @@
--- |
--- Provides the ability to sort modules based on module dependencies
---
+-- | Provides the ability to sort modules based on module dependencies
 module Language.PureScript.ModuleDependencies
   ( sortModules
   , ModuleGraph
   ) where
 
 import Prelude.Compat
 
+import Control.Monad (forM_, when)
 import Control.Monad.Error.Class (MonadError(..))
-
 import Data.Graph
 import Data.List (nub)
 import Data.Maybe (fromMaybe)
-
 import Language.PureScript.AST
 import Language.PureScript.Crash
 import Language.PureScript.Errors
 import Language.PureScript.Names
-import Language.PureScript.Types
 
 -- | A list of modules with their transitive dependencies
 type ModuleGraph = [(ModuleName, [ModuleName])]
 
 -- | Sort a collection of modules based on module dependencies.
 --
 -- Reports an error if the module graph contains a cycle.
---
-sortModules :: (MonadError MultipleErrors m) => [Module] -> m ([Module], ModuleGraph)
+sortModules
+  :: forall m
+   . MonadError MultipleErrors m
+  => [Module]
+  -> m ([Module], ModuleGraph)
 sortModules ms = do
-  let verts = map goModule ms
-  ms' <- mapM toModule $ stronglyConnComp verts
-  let (graph, fromVertex, toVertex) = graphFromEdges verts
-      moduleGraph = do (_, mn, _) <- verts
-                       let v       = fromMaybe (internalError "sortModules: vertex not found") (toVertex mn)
-                           deps    = reachable graph v
-                           toKey i = case fromVertex i of (_, key, _) -> key
-                       return (mn, filter (/= mn) (map toKey deps))
-  return (ms', moduleGraph)
+    let mns = map getModuleName ms
+    verts <- mapM (toGraphNode mns) ms
+    ms' <- mapM toModule $ stronglyConnComp verts
+    let (graph, fromVertex, toVertex) = graphFromEdges verts
+        moduleGraph = do (_, mn, _) <- verts
+                         let v       = fromMaybe (internalError "sortModules: vertex not found") (toVertex mn)
+                             deps    = reachable graph v
+                             toKey i = case fromVertex i of (_, key, _) -> key
+                         return (mn, filter (/= mn) (map toKey deps))
+    return (ms', moduleGraph)
   where
-  goModule :: Module -> (Module, ModuleName, [ModuleName])
-  goModule m@(Module _ _ _ ds _) =
-    let ams = concatMap extractQualAs ds
-    in (m, getModuleName m, nub (concatMap (usedModules ams) ds))
+    toGraphNode :: [ModuleName] -> Module -> m (Module, ModuleName, [ModuleName])
+    toGraphNode mns m@(Module _ _ mn ds _) = do
+      let deps = nub (concatMap usedModules ds)
```

Comment:
This could be very expensive for lists with 30 modulenames that start with: "Slamdata.Editor". ordNub is just better.